src/storage/src/render/upsert.rs

```diff
@@ -206,23 +227,36 @@ where
                     // From the prefix that can be emitted we can deduplicate based on (ts, key) in
                     // order to only process the command with the maximum order within the (ts,
                     // key) group. This is achieved by wrapping order in `Reverse(order)` above.
-                    let mut commands = stash.drain(..idx).dedup_by(|a, b| {
-                        let ((a_ts, a_key, _, _), (b_ts, b_key, _, _)) = (a, b);
-                        a_ts == b_ts && a_key == b_key
-                    });
-
-                    while let Some((ts, key, _, value)) = commands.next() {
-                        match value {
-                            Some(value) => {
-                                if let Some(old_value) = state.insert(key, value.clone()) {
-                                    output_updates.push((old_value, ts.clone(), -1));
-                                }
-                                output_updates.push((value, ts, 1));
+                    let mut commands = stash
+                        .drain(..idx)
+                        .dedup_by(|a, b| {
+                            let ((a_ts, a_key, _, _), (b_ts, b_key, _, _)) = (a, b);
+                            a_ts == b_ts && a_key == b_key
+                        })
+                        .peekable();
+
+                    // Only upsert (and update stats) if there are actually new values
+                    // to upsert.
+                    if commands.peek().is_some() {
+                        for UpsertStateResult {
+                            value,
+                            previous_value,
+                            ts,
+                        } in state
+                            .upsert_batch(commands.map(|(ts, key, _, value)| UpsertStateKV {
```

Comment:
Ah I see what you meant. In my PR I was getting the old state comparing and then inserting [link](https://github.com/MaterializeInc/materialize/pull/18567/files#diff-183524b2b5b9ca2e5b08857bb7e8913c1efebda1d217fc902d24507f30e283feR272-R294)

Reply:
Yeah, we might need to move that into the trait! will discuss this afternoon!